File: 20-valid-parentheses/valid-parentheses.cpp
class Solution {
public:
    bool isValid(string s) {
        if(s.size()%2 !=0) return false;
        std::stack<char> st;
        for(auto ch : s){
            if(ch == '(' || ch == '{' || ch == '[') st.push(ch);
            else {
                if(st.empty()) return false;
                if((ch == ')' && st.top()!='(') || (ch == '}' && st.top()!='{') || (ch == ']' && st.top()!='[')) return false;
                else st.pop();
            }
        }
        if(!st.empty()) return false;
        return true;
    }
};

File: 20-valid-parentheses/valid-parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        if len(s)%2 !=0 : return False
        st = []
        for ch in s:
            if ch=='(' or ch=='{' or ch=='[' : st.append(ch)
            else:
                if len(st)==0: return False
                elif ch == ')' and st[-1]!='(': return False
                elif ch == '}' and st[-1]!='{': return False
                elif ch == ']' and st[-1]!='[': return False
                else: st.pop()
        
        if len(st)!=0: return False
        return True
